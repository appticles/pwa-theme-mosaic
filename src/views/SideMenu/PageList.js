import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { List, Header, Icon, Modal, Image } from 'semantic-ui-react';

import './style.css';

export default class PageList extends Component {
  state = { parentIds: [], pages: [], parentPageTitle: ['Go to'], goBack: false };

  componentWillReceiveProps() {
    this.setState({ parentIds: [...this.state.parentIds, 0] });
    // resets state when menu reopens
    this.props.visible === true && this.setState({ parentIds: [], pages: [], parentPageTitle: ['Go to'], goBack: false });
  }
  setParentsList(parentId, pages, parentPageTitle) {
    this.setState({
      parentIds: [...this.state.parentIds, parentId],
      pages: [...this.state.pages, pages],
      parentPageTitle: [...this.state.parentPageTitle, parentPageTitle],
      goBack: false,
    });
  }
  setParentsBackList() {
    this.state.parentIds.pop();
    const poppedParentIds = this.state.parentIds;
    this.state.parentPageTitle.pop();
    const poppedPageTitles = this.state.parentPageTitle;
    this.setState({ parentId: poppedParentIds, parentsPageTitle: poppedPageTitles, goBack: true });
  }

  render() {
    const checkParent = this.state.parentIds[this.state.parentIds.length - 1];
    let parents;
    this.state.goBack === false ? (parents = this.props.pages.filter(page => page.parent === checkParent)) : (parents = this.state.pages.pop());

    function getImage(sourceImg) {
      let imageSource;
      if (sourceImg) {
        imageSource = sourceImg[0].source_url;
        return imageSource;
      }
      return imageSource;
    }

    return (
      <List divided relaxed>
        {checkParent === 0 ? (
          <Header>{this.state.parentPageTitle[this.state.parentPageTitle.length - 1]}</Header>
        ) : (
          <Header>
            <Icon name="caret left" onClick={e => this.setParentsBackList()} />
            {this.state.parentPageTitle[this.state.parentPageTitle.length - 1]}
          </Header>
        )}
        {parents.map(page => (
          <List.Item key={Math.random()}>
            <List.Icon name="linkify" size="large" verticalAlign="middle" />
            <List.Content>
              <Modal closeIcon trigger={<List.Header as="a">{page.title.rendered}</List.Header>}>
                <Modal.Header dangerouslySetInnerHTML={{ __html: page.title.rendered }} />
                <Image src={getImage(page._embedded['wp:featuredmedia'])} size="medium" />
                <Modal.Content dangerouslySetInnerHTML={{ __html: page.content.rendered }} />
                <div className="go-to">
                  <Link to={'/page/' + page.slug + '/' + page.id}>More</Link>
                </div>
              </Modal>
            </List.Content>
            {page.children.length > 0 ? (
              <List.Icon
                link
                name="triangle right"
                size="large"
                verticalAlign="middle"
                onClick={e => this.setParentsList(page.id, parents, page.title.rendered)}
              />
            ) : (
              ''
            )}
          </List.Item>
        ))}
      </List>
    );
  }
}
