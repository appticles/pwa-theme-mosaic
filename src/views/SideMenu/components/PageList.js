import React, { Component } from 'react';
import { List, Header, Icon } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import config from 'react-global-configuration';
import _ from 'lodash';

import { pagePropType } from '../reducer';

import '../style.css';

export default class PageList extends Component {
  constructor(props) {
    super(props);

    this.goBack = this.goBack.bind(this);
    this.openSubmenu = this.openSubmenu.bind(this);

    this.state = {
      parentId: 0,
      parentPageTitle: [],
      visible: true,
      pages: [],
    };
  }

  /**
   * Add a children property to each page - does/not have kids.
   * @param {Object} previousProps
   */
  componentDidUpdate(previousProps) {
    const pages = this.props.pages.map((item) => {
      const children = previousProps.pages.filter(child => child.parent === item.id);
      return Object.assign(item, { children: children.length > 0 });
    });

    if (!_.isEqual(this.state.pages, pages)) {
      this.setState({ pages });
    }
  }

  filterPages() {
    return this.state.pages.filter(page => page.parent === this.state.parentId);
  }

  openSubmenu(parentId, parentPageTitle) {
    this.setState({ parentId, parentPageTitle: [...this.state.parentPageTitle, parentPageTitle] });
  }

  goBack() {
    const parentPage = this.props.pages.filter(page => page.id === this.state.parentId);
    const titles = this.state.parentPageTitle.slice();
    titles.pop();
    this.setState({ parentId: parentPage[0].parent, parentPageTitle: titles });
  }

  render() {
    const pages = this.filterPages();

    return (
      <List divided relaxed>
        {this.state.parentId === 0 ? (
          <Header as="h2">{this.props.text && this.props.text.GO_TO}</Header>
        ) : (
          <Header as="h2">
            <Icon name="caret left" onClick={this.goBack} />
            {this.state.parentPageTitle[this.state.parentPageTitle.length - 1]}
          </Header>
        )}
        {pages.map(page => (
          <List.Item key={Math.random()}>
            <List.Icon name="linkify" size="large" verticalAlign="middle" />
            <List.Content>
              <Link to={'/page/' + page.slug + '/' + page.id}>
                <List.Header as="h3">{page.title.rendered}</List.Header>
              </Link>
            </List.Content>
            {page.children ? (
              <List.Icon
                link
                name="triangle right"
                size="large"
                verticalAlign="middle"
                onClick={() => this.openSubmenu(page.id, page.title.rendered)}
              />
            ) : null}
          </List.Item>
        ))}
        {this.state.parentId === 0 && config.get('websiteUrl') ? (
          <List.Item>
            <List.Icon name="linkify" size="large" verticalAlign="middle" />
            <List.Content>
              <a href={config.get('websiteUrl')}>
                <List.Header as="h3">
                  {this.props.text && this.props.text.VISIT_WEBSITE}
                </List.Header>
              </a>
            </List.Content>
          </List.Item>
        ) : null}
      </List>
    );
  }
}

PageList.defaultProps = {
  text: {
    GO_TO: 'Go to',
    VISIT_WEBSITE: 'Visit website',
  },
};
PageList.propTypes = {
  pages: PropTypes.arrayOf(pagePropType).isRequired,
  text: PropTypes.shape({
    GO_TO: PropTypes.string,
    VISIT_WEBSITE: PropTypes.string,
  }),
};
