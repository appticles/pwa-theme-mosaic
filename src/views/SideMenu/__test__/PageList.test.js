import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import PageList from '../PageList';

describe('Page list', () => {
  it('renders page list in side menu', () => {
    const pages = [
      {
        id: 53,
        parent: 5,
        title: {
          rendered: 'Page Title',
        },
        slug: 'page-slug',
        content: {
          rendered: 'Page content',
        },
        link: 'page-link',
        author: 68,
        comment_status: 'open',
        featured_media: 36,
        status: 'status',
      },
    ];
    const text = {
      GO_TO: 'Go to',
      VISIT_WEBSITE: 'Visit website',
    };
    const jsonSettings = {
      websiteUrl: 'some-website-url',
    };
    const output = shallow(<PageList pages={pages} text={text} websiteUrl={jsonSettings} />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
