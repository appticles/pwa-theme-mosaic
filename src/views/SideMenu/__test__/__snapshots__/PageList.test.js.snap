// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page list renders page list in side menu 1`] = `
<List
  divided={true}
  relaxed={true}
>
  <Header>
    Go to
  </Header>
</List>
`;
